### PR TITLE
opt, schedulers: consider placement rules when verify region healthy

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### What problem does this PR solve? 
continue merging placement rules.

### What is changed and how it works?
- replace `core.RegionHealth` with `opt.RegionHealth`. When placement rules is enabled, it allows learner peers.
- replace `core.RegionHealthAllowPending` with `opt.RegionHealthAllowPending`. When placement rules is enabled, it allows learner peers.
- replace `len(region.GetPeers()) == cluster.GetMaxReplicas()` with `opt.IsRegionReplicated`. When placement rules is enabled, it checks if peers can fit rules.

### Check List 
Tests 
 - Unit test

### CODE REVIEW

File: server/core/region_option.go [hunk 1]
```diff
@@ -55,7 +41,15 @@ func WithPendingPeers(pengdingPeers []*metapb.Peer) RegionCreateOption {
 // WithLearners sets the learners for the region.
 func WithLearners(learners []*metapb.Peer) RegionCreateOption {
 	return func(region *RegionInfo) {
-		region.learners = learners
+		peers := region.meta.GetPeers()
+		for i := range peers {
+			for _, l := range learners {
```
Comment (shafreeck) [inline server/core/region_option.go]: Just mention it, why not try to keep the same style of the loop like `for j := range learners`, then use `learners[j]` later instead.

Comment (disksing) [reply server/core/region_option.go]: only because `l` is shorter than `learners[j]`.

File: server/schedule/opt/healthy.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opt
+
+import "github.com/pingcap/pd/server/core"
+
+// IsRegionHealthy checks if a region is healthy for scheduling. It requires the
+// region does not have any down or pending peers, and does not have any learner
+// peers when placement rules is disabled.
```
Comment (lhy1024) [inline server/schedule/opt/healthy.go]: I have a bit confused what means `does not have any learner peers when placement rules is disabled.`

File: server/schedulers/hot_region.go [hunk 3]
```diff
@@ -303,12 +303,12 @@ func (h *balanceHotRegionsScheduler) balanceByPeer(cluster opt.Cluster, storesSt
 			continue
 		}
 
-		if isRegionUnhealthy(srcRegion) {
+		if !opt.IsRegionHealthyAllowPending(cluster, srcRegion) {
```
[the remark this answers is not in the document]
Comment (disksing) [reply server/schedulers/hot_region.go]: Any suggestion? `IsHealthyAllowPending`?

File: server/schedule/opt/healthy.go (hunk 2 above)
Comment (disksing) [reply server/schedule/opt/healthy.go]: `And when placement rules feature is disabled, it requires the region does not have any learner peer`. Looks better?

**Pull request merged**